Dfe.Academies.External.Web/Pages/WhatIsYourRole.cshtml.cs

```diff
@@ -1,12 +1,104 @@
+
+public class WhatIsYourRoleModel : BasePageModel
 {
-    public class WhatIsYourRoleModel : PageModel
+    private readonly ILogger<WhatIsYourRoleModel> _logger;
+    private readonly IConversionApplicationCreationService _academisationCreationService;
+    private ConversionApplication _draftConversionApplication;
+    private const string NextStepPage = "/ApplicationOverview";
+
+    public WhatIsYourRoleModel(ILogger<WhatIsYourRoleModel> logger,
+                                IConversionApplicationCreationService academisationCreationService)
+    {
+        _logger = logger;
+        _academisationCreationService = academisationCreationService;
+    }
+
+    [BindProperty]
+    [RequiredEnum(ErrorMessage = "You must give your role at the school")]
+    public SchoolRoles SchoolRole { get; set; }
+
+    [BindProperty]
+    public string? OtherRoleNotListed { get; set; }
+
+    public bool OtherRoleError
+    {
+        get
+        {
+            if (!ModelState.IsValid && ModelState.Keys.Contains("OtherRoleNotEntered"))
+            {
+                return true;
+            }
+
+            return false;
+        }
+    }
+
+    public async Task OnGetAsync()
+    {
+        //// on load - grab draft application from temp
+        _draftConversionApplication = TempDataHelper.GetSerialisedValue<ConversionApplication>("draftConversionApplication", TempData) ?? new ConversionApplication();
+    }
+
+    public async Task<IActionResult> OnPostAsync()
+    {
+        if (!ModelState.IsValid)
+        {
+            // error messages component consumes ViewData["Errors"]
+            PopulateValidationMessages();
+            return Page();
+        }
+
+        if (SchoolRole == SchoolRoles.Other && string.IsNullOrWhiteSpace(OtherRoleNotListed))
+        {
+            ModelState.AddModelError("OtherRoleNotEntered", "You must give your role at the school");
+            PopulateValidationMessages();
+            return Page();
+        }
+
+        //// grab draft application from temp
+        _draftConversionApplication = TempDataHelper.GetSerialisedValue<ConversionApplication>("draftConversionApplication", TempData) ?? new ConversionApplication();
+
+        try
+        {
+            _draftConversionApplication.SchoolRole = SchoolRole;
+            _draftConversionApplication.OtherRoleNotListed = OtherRoleNotListed;
+
+            await _academisationCreationService.UpdateDraftApplication(_draftConversionApplication);
+
+            // update temp store for next step
+            TempDataHelper.StoreSerialisedValue("draftConversionApplication", TempData, _draftConversionApplication);
+
+            return RedirectToPage(NextStepPage);
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError("Application::WhatIsYourRoleModel::OnPostAsync::Exception - {Message}", ex.Message);
+            return Page();
+        }
+    }
+
+    public override void PopulateValidationMessages()
```

Comment:
I feel like this validation stuff could be a bit more elegant but I'm not sure exactly how and it's probably not a priority for now. If I get an hour at some point I might have a play around and see what we could do.

Reply:
sure, sure, sure